### PR TITLE
Algolia DocSearch Integration 

### PR DESCRIPTION
**Pull request by ulivz**
1. Try to fit the style from vue.js official website.
2. Need to clean config and add document before merged.

> A point to be considered: Since the latest docsearch.js render the distinct default style, so there are many css resets to make it looks good.

> Preview: https://deploy-preview-201--vuepress.netlify.com/

Basically completed, welcome to test and feel free to give your comments.

### CODE REVIEW

File: docs/.vuepress/config.js [hunk 1]
```diff
@@ -28,6 +28,12 @@ module.exports = {
     repo: 'vuejs/vuepress',
     editLinks: true,
     docsDir: 'docs',
+    algolia: {
+      appId: 'BH4D9OD16A',
```
Comment (bencodezen) [inline docs/.vuepress/config.js]: Should the appId and apiKey be committed as part of the repo?

Comment (ulivz) [reply docs/.vuepress/config.js]: Hmmm... At thePR summary, I have said:

> Need to clean config ... before merged.

**Comment by ulivz** [comment]
Seeing so many css reset, I decided to fallback to docSearch@1.5.0 that the official website uses.

**Comment by yyx990803** [comment]
Oh nvm, if you are still working on it please let me know when you finish.

**Comment by yyx990803** [comment]
I think we should use the latest version of docsearch if possible. CSS resets are not that much of a problem.

**Comment by yyx990803** [comment]
BTW we need to limit the max width of the search box on mobile:

screen shot 2018-04-23 at 11 37 52 am

**Comment by ulivz** [comment]
OK, please wait for a moment. I need to limit the width of the mobile phone and test again.

**Comment by yyx990803** [comment]
Can you add documentation for the `algolila` option please

File: lib/default-theme/Navbar.vue [hunk 2]
```diff
@@ -12,7 +12,7 @@
       </span>
     </router-link>
     <div class="links">
-      <AlgoliaSearchBox v-if="isAlgoliaSearch" :options="$site.themeConfig.algolia"/>
+      <AlgoliaSearchBox v-if="isAlgoliaSearch" :options="algolia"/>
       <SearchBox v-if="isSearch"/>
```
Comment (yyx990803) [inline lib/default-theme/Navbar.vue]: Can be `v-else-if="$site.themeConfig.search !== false"` here

File: docs/zh/config/README.md [hunk 3]
```diff
@@ -97,6 +97,24 @@ module.exports = {
 当然，仅仅只在你的网站部署后能用 SSL 的时候开启它，因为 service worker 只能在 HTTPs 的链接下注册。
 :::
 
+### algolia
+
+- 类型: `Object`
+- 默认值: `undefined`
+
+使用 `algolia` 选项可以让你禁用掉默认的基于 headers 的搜索，从而使用 [algolia docsearch](https://github.com/algolia/docsearch)。为了使其生效，你必须提供至少 `apiKey` 和 `indexName` 这两个选项：
```
Comment (yyx990803) [inline docs/zh/config/README.md]: ...可以让你用 algolia search 取代默认的...

**Comment by ulivz** [comment]
@yyx990803 

- Implemented locale-scoped docsearch at 702d676
- document for that:  ea8aa21 and a95aa89

File: docs/zh/config/README.md [hunk 4]
```diff
@@ -97,6 +97,24 @@ module.exports = {
...
+```js
+module.exports = {
+  algolia: {
+    apiKey: '<API_KEY>',
+    indexName: '<INDEX_NAME>'
+  }
+}
+```
+
+其他可用的选项可以参考 [docsearch options](https://github.com/algolia/docsearch#docsearch-options)。
```
Comment (yyx990803) [inline docs/zh/config/README.md]: "请"参考

File: lib/default-theme/Navbar.vue [hunk 5]
```diff
@@ -12,19 +12,33 @@
       </span>
     </router-link>
     <div class="links">
-      <SearchBox v-if="$site.themeConfig.search !== false"/>
+      <AlgoliaSearchBox v-if="isAlgoliaSearch" :options="algolia"/>
+      <SearchBox v-if="isSearch"/>
```
Comment (yyx990803) [inline lib/default-theme/Navbar.vue]: 可以用 `v-else-if="$site.themeConfig.search !== false"`

File: lib/default-theme/AlgoliaSearchBox.vue [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+export default {
+  props: ['options'],
+  mounted () {
+    this.initialize()
+  },
+  methods: {
+    initialize () {
+      Promise.all([
+        import(/* webpackChunkName: "docsearch" */ 'docsearch.js/dist/cdn/docsearch.min.js'),
+        import(/* webpackChunkName: "docsearch" */ 'docsearch.js/dist/cdn/docsearch.min.css')
```
Comment (yyx990803) [inline lib/default-theme/AlgoliaSearchBox.vue]: 现在的 webpack 配置会把所有 css 都抽取出来，即使用 dynamic import 也没用。这样的话不论用不用 algolia 都会加载这些 css。有一个办法是在 createBaseConfig 的时候如果用户没有使用 algolia 则直接把 AlgoliaSearchBox alias 到一个空 module, 这样可以彻底避免把这个组件打包进去。

Comment (ulivz) [reply lib/default-theme/AlgoliaSearchBox.vue]: 厉害了，我还没注意这个

**Comment by jerome2710** [comment]
It looks like `appId` is also required, otherwise it will fallback on the default appId from the docsearch.js NPM-package which throws an error. Could this be added to the documentation as well please?

Also, could some description be added on how the content is inserted into Algolia? I am not getting any search results right now, do I need to add records to the index myself?

**Comment by jerome2710** [comment]
Hi @bencodezen, thank you for your reply. No, I was unaware Algolia needed to scrape the docs-domain to index the content. I just submitted the site! Maybe add this to the documentation as well?

**Comment by bencodezen** [comment]
@jerome2710 Here's the PR for the docs update: https://github.com/vuejs/vuepress/pull/634/commits/6082bc159db8a026ac73d53515fe11eeb34e0a2c

**Comment by jerome2710** [comment]
@bencodezen Very useful, thank you! Should we mention the `appId` here as well?

**Comment by bencodezen** [comment]
@jerome2710 I'm currently using Algolia for VueMeetups.org and did not configure an appId. Could you send me a gist or something to show what you're talking about?

**Comment by jerome2710** [comment]
@bencodezen Interesting to hear that it does work for you. I am using v0.10.2 of Vuepress and that includes the Algolia npm-package of docsearch.js with v2.5.2. As you can see [here](https://github.com/algolia/docsearch/blob/master/src/lib/DocSearch.js#L36), the constructor falls back to an default appId. This causes an appId/apiKey mismatch for me if I omit it in the config.js of Vuepress.

**Pull request merged**